### PR TITLE
Test case to check if receiver blocks on empty channel, and resumes on close.

### PR DESCRIPTION
**Pull request by kavyasrinet**

### CODE REVIEW

File: paddle/framework/channel_test.cc [hunk 1]
```diff
@@ -82,6 +82,33 @@ TEST(Channel, ConcurrentSendNonConcurrentReceiveWithSufficientBufferSize) {
   delete ch;
 }
 
+void ReceiveEmptyChannelTest(Channel* ch) {
+  size_t current = 5;
+  std::thread t([&]() {
+    // Read from channel
+    int out;
+    ch->Receive(&out); // should block since channel is empty
+    current = 10;
```
Comment (wangkuiyi) [inline paddle/framework/channel_test.cc]: 10 is a magical number

File: paddle/framework/channel_test.cc [hunk 2]
```diff
@@ -82,6 +82,33 @@ TEST(Channel, ConcurrentSendNonConcurrentReceiveWithSufficientBufferSize) {
   delete ch;
 }
 
+void ReceiveEmptyChannelTest(Channel* ch) {
+  size_t current = 5;
```
Comment (wangkuiyi) [inline paddle/framework/channel_test.cc]: 5 is a magical number

File: paddle/framework/channel_test.cc [hunk 3]
```diff
@@ -82,6 +82,33 @@ TEST(Channel, ConcurrentSendNonConcurrentReceiveWithSufficientBufferSize) {
...
+  std::thread t([&]() {
+    // Read from channel
```
Comment (wangkuiyi) [inline paddle/framework/channel_test.cc]: unnecessary comment

File: paddle/framework/channel_test.cc [hunk 4]
```diff
@@ -82,6 +82,33 @@ TEST(Channel, ConcurrentSendNonConcurrentReceiveWithSufficientBufferSize) {
   delete ch;
 }
 
+void ReceiveEmptyChannelTest(Channel* ch) {
```
Comment (wangkuiyi) [inline paddle/framework/channel_test.cc]: The  prefix Test is a convention of unit test names.  This function is not a unit test; a better name could be `ReceiveFromEmptyChannelShouldBlock`.

File: paddle/framework/channel_test.cc (hunk 1 above)
Comment (kavyasrinet) [reply paddle/framework/channel_test.cc]: Thank you for the review. 
I just realized that we already checked in the test cases : [BufferedChannelCloseUnblocksReceiversTest](https://github.com/PaddlePaddle/Paddle/blob/develop/paddle/framework/channel_test.cc#L108) and [UnbufferedChannelCloseUnblocksReceiversTest](https://github.com/PaddlePaddle/Paddle/blob/develop/paddle/framework/channel_test.cc#L204) that perform these operations already.
In that case, this PR would be redundant and maybe I can close it.

**Comment by wangkuiyi** [comment]
A rewrite is as follows:

```c++
void ReceiveFromEmptyChannelShouldBlockAndCloseUnblocksIt(Channel *ch) {
  bool received = false;
  std::thread t([&]() {
    int out;
    ch->Receive(&out);  // should block since channel is empty
    received = true;
  });
  std::this_thread::sleep_for(std::chrono::milliseconds(100));  // wait 0.1 sec
  EXPECT_EQ(received,
            false);  // the receiver should be blocked since channel is empty

  CloseChannel(ch);
  t.join();
  EXPECT_EQ(received,
            true);  // when we close the channel, the receiver should unblock
  delete ch;
}

TEST(Channel, ReceiveFromEmptyBufferedChannelShouldBlockAndCloseUnblocksIt) {
  auto ch = MakeChannel<size_t>(3);  // buffered
  ReceiveFromEmptyChannelShouldBlockAndCloseUnblocksIt(ch);
}

TEST(Channel, ReceiveFromEmptyUnBufferedChannelShouldBlockAndCloseUnblocksIt) {
  auto ch = MakeChannel<size_t>(0);  // unbuffered
  ReceiveEmptyChannelTest(ch);
}
```

**Pull request closed without merging**